### PR TITLE
STK-1411-fix-datetime-comparison

### PR DESCRIPTION
**Pull request by garethj2**
Fixes the date time comparison in haystack core. This now deals properly with timezone offsets.

HDict has also been updated.

### CODE REVIEW

File: spec/core/HDict.spec.ts [hunk 1]
```diff
@@ -844,25 +844,41 @@ describe('HDict', function (): void {
 			expect(dict2.isNewer(dict1)).toBe(false)
 		})
 
-		it('returns false if mod is not defined and dicts content is the same', function (): void {
+		it('returns true if mod is not defined in for first dict', function (): void {
```
Comment (riccardolederj2) [inline spec/core/HDict.spec.ts:847]:
```suggestion
		it('returns true if mod is not defined in second dict', function (): void {
```

File: spec/core/HDict.spec.ts [hunk 2]
```diff
@@ -844,25 +844,41 @@ describe('HDict', function (): void {
...
 			dict2 = new HDict({
 				foo: 'foo',
 			})
 
-			expect(dict1.isNewer(dict2)).toBe(false)
+			expect(dict1.isNewer(dict2)).toBe(true)
 		})
 
-		it('returns true if mod is not defined and dict1 content differs', function (): void {
+		it('returns true if mod is not defined in for second dict', function (): void {
```
Comment (riccardolederj2) [inline spec/core/HDict.spec.ts:860]:
```suggestion
		it('returns true if mod is not defined in first dict', function (): void {
```

File: src/core/HDict.ts [hunk 3]
```diff
@@ -987,17 +987,17 @@ export class HDict implements HVal, Iterable<HValRow> {
...
 	 * @return True if this dict is newer than the other dict
 	 */
 	public isNewer(dict: HDict): boolean {
-		const modA = this.get<HDateTime>('mod')?.value || ''
-		const modB = dict.get<HDateTime>('mod')?.value || ''
+		const a = this.get<HDateTime>('mod')
+		const b = dict.get<HDateTime>('mod')
 
-		return modA > modB || (modA === modB && !this.equals(dict))
+		return !(a && b) || a.compareTo(b) >= 0
```
Comment (riccardolederj2) [inline src/core/HDict.ts:1001]: What is the reason behind removing the dict `equals` comparison?
In this way this method seems to be more of a `isNewerOrEquallyOld`..

**Review by riccardolederj2 — APPROVED** [review]
Looks good to me and Greta, the only doubt is concerning the removal of the equals check.

File: src/core/HDict.ts (hunk 3 above)
Comment (garethj2) [reply src/core/HDict.ts:1001]: When the `compareTo` is called, it's already done the check to ensure `a` and `b` are both real `HDateTime` values.

I wanted to give the benefit of the doubt and drop the `equals` check and hence I thought equals would be better.

You're right though. The name should change to something like `isNewerOrSame()` or something like that. What do you think?

Comment (garethj2) [reply src/core/HDict.ts:1001]: I just can't think of a good name here.

**Review by rracariu — APPROVED** [review]

**Review by riccardolederj2 — APPROVED** [review]
LGTM

**Comment by garethj2** [comment]
The method didn't seem right so I've stripped it back. 

The features required for using this with watches should go in haystack-nclient and not here.

**Pull request merged**